File: scripts/inventory_calculations.py
#!/usr/bin/env python3
"""
재고 관리 계산 함수 (Python)
- 실시간 가용 재고 (물류센터 마감 시차 보정)
- 재고 소진일(Run-out date) 예측
- 쿠팡/일반 매출 분리 분석

1년+ 시계열 데이터 기반 예측·통제용
"""

from dataclasses import dataclass
from datetime import date, datetime, timedelta
from typing import Literal, Optional

SalesChannel = Literal["coupang", "general"]


@dataclass
class LogisticsCutoff:
    """물류센터 마감 시각 (KST)"""
    channel: SalesChannel
    hour: int  # 0-23
    minute: int = 0


DEFAULT_CUTOFF = [
    LogisticsCutoff("coupang", 18, 0),   # 쿠팡: 18:00 마감
    LogisticsCutoff("general", 23, 59),  # 일반: 23:59 마감
]


def is_before_cutoff(
    channel: SalesChannel,
    now: Optional[datetime] = None,
    cutoffs: Optional[list[LogisticsCutoff]] = None,
) -> bool:
    """현재 시각이 해당 채널의 당일 마감 전인지"""
    now = now or datetime.now()
    cutoffs = cutoffs or DEFAULT_CUTOFF
    cfg = next((c for c in cutoffs if c.channel == channel), None)
    if not cfg:
        return True
    cutoff_minutes = cfg.hour * 60 + cfg.minute
    now_minutes = now.hour * 60 + now.minute
    return now_minutes < cutoff_minutes


def compute_real_time_available_stock(
    current_stock: int | float,
    today_out_coupang: int | float,
    today_out_general: int | float,
    now: Optional[datetime] = None,
) -> float:
    """
    실시간 가용 재고 (물류센터 마감 시차 보정)
    
    마감 전: 해당 채널의 당일 출고분이 아직 반영 안 됨 → 가용 = 현재고 - 미반영 출고
    """
    now = now or datetime.now()
    pending = 0.0
    if is_before_cutoff("coupang", now):
        pending += today_out_coupang
    if is_before_cutoff("general", now):
        pending += today_out_general
    return max(0, current_stock - pending)


def get_average_daily_outbound(
    outbound_records: list[dict],  # [{"date": "2026-01-01", "quantity": 100, "item_id": "mask", "sales_channel": "coupang"}]
    item_id: str,
    days: int = 30,
    by_product: bool = False,
    product_code_key: str = "product_code",
) -> float:
    """최근 N일 평균 일일 출고량 (실제 거래일 기준)"""
    cutoff = date.today() - timedelta(days=days)
    daily_total: dict[str, float] = {}
    total_qty = 0.0

    for r in outbound_records:
        try:
            rdate = r["date"] if isinstance(r["date"], date) else datetime.strptime(r["date"], "%Y-%m-%d").date()
        except (ValueError, TypeError):
            continue
        if rdate < cutoff:
            continue
        key = r.get(product_code_key, item_id) if by_product else item_id
        match = (r.get(product_code_key) == item_id) if by_product else (r.get("item_id") == item_id)
        if not match:
            continue
        dstr = rdate.isoformat()
        daily_total[dstr] = daily_total.get(dstr, 0) + r.get("quantity", 0)
        total_qty += r.get("quantity", 0)

    day_count = len(daily_total)
    return total_qty / day_count if day_count > 0 else 0.0


def predict_run_out_date(
    current_stock: int | float,
    avg_daily_out: float,
    from_date: Optional[date] = None,
) -> tuple[Optional[str], int, bool]:
    """
    재고 소진일 예측
    Returns: (예상 소진일 YYYY-MM-DD, 잔여일수, 무한대 여부)
    """
    from_date = from_date or date.today()
    if current_stock <= 0:
        return from_date.isoformat(), 0, False
    if avg_daily_out <= 0:
        return None, 999, True
    days_left = int(current_stock / avg_daily_out)
    run_out = from_date + timedelta(days=days_left)
    return run_out.isoformat(), days_left, False


def predict_all_run_out_dates(
    stock: dict[str, int | float],
    outbound_records: list[dict],
    item_names: dict[str, str],
    lookback_days: int = 30,
) -> list[dict]:
    """품목별 소진일 예측 결과"""
    results = []
    for item_id, current in stock.items():
        avg = get_average_daily_outbound(outbound_records, item_id, lookback_days)
        date_str, days_left, is_infinite = predict_run_out_date(current, avg)
        results.append({
            "item_id": item_id,
            "item_name": item_names.get(item_id, item_id),
            "current_stock": current,
            "avg_daily_out": round(avg, 1),
            "run_out_date": date_str,
            "days_left": 999 if is_infinite else days_left,
            "is_infinite": is_infinite,
            "is_urgent": not is_infinite and 0 <= days_left <= 7,
        })
    return sorted(results, key=lambda x: (x["is_infinite"], x["days_left"]))


# --- Excel 연동 예시 (pandas) ---
def load_outbound_from_excel(path: str, sheet: str = "출고") -> list[dict]:
    """생산수불현황 Excel 출고 시트 로드"""
    import pandas as pd
    df = pd.read_excel(path, sheet_name=sheet, header=None)
    # 헤더 탐색 (품목코드, 제품명, 품목구분, 수량, 매출구분, 출고일자)
    # 실제 컬럼 인덱스는 Excel 구조에 맞게 조정
    records = []
    for _, row in df.iterrows():
        # row 처리 로직 (실제 구조에 맞게 수정)
        records.append({
            "date": str(row.get("출고일자", "")),
            "quantity": int(row.get("수량", 0) or 0),
            "item_id": str(row.get("품목구분", "")),
            "product_code": str(row.get("품목코드", "")),
            "sales_channel": "coupang" if "쿠팡" in str(row.get("매출구분", "")) else "general",
        })
    return records


if __name__ == "__main__":
    # 사용 예시
    stock = {"mask": 5000, "capsule": 10000, "fabric": 8000}
    outbound = [
        {"date": "2026-03-01", "quantity": 200, "item_id": "mask"},
        {"date": "2026-03-02", "quantity": 180, "item_id": "mask"},
        # ... 30일치
    ]
    names = {"mask": "마스크", "capsule": "캡슐세제", "fabric": "섬유유연제"}
    preds = predict_all_run_out_dates(stock, outbound, names)
    for p in preds:
        print(f"{p['item_name']}: {p['run_out_date']} ({p['days_left']}일)")

File: src/lib/inventoryCalculations.ts
/**
 * 재고 관리 계산 함수
 * - 실시간 가용 재고 (물류센터 마감 시차 보정)
 * - 재고 소진일(Run-out date) 예측
 * - 쿠팡/일반 매출 분리 분석
 */

import type { Transaction } from "./types";
import { mapGroupToItemId } from "./unifiedImport";

export type SalesChannel = "coupang" | "general";

/** 물류센터 마감 시각 (KST 기준) */
export interface LogisticsCutoff {
  channel: SalesChannel;
  hour: number;   // 0-23
  minute: number; // 0-59
}

const DEFAULT_CUTOFF: LogisticsCutoff[] = [
  { channel: "coupang", hour: 18, minute: 0 },  // 쿠팡: 18:00 마감
  { channel: "general", hour: 23, minute: 59 }, // 일반: 23:59 마감
];

/** Transaction에 매출구분(쿠팡/일반) 필드 확장 */
export interface TransactionWithChannel extends Transaction {
  salesChannel?: SalesChannel;
}

/**
 * 현재 시각이 해당 채널의 "당일 마감 전"인지 판단
 * @param channel 쿠팡 | 일반
 * @param now 기준 시각 (Date)
 * @param cutoffs 마감 시각 설정 (없으면 기본값)
 */
export function isBeforeCutoff(
  channel: SalesChannel,
  now: Date = new Date(),
  cutoffs: LogisticsCutoff[] = DEFAULT_CUTOFF
): boolean {
  const cfg = cutoffs.find((c) => c.channel === channel) ?? {
    channel,
    hour: 23,
    minute: 59,
  };
  const cutoffMinutes = cfg.hour * 60 + cfg.minute;
  const nowMinutes = now.getHours() * 60 + now.getMinutes();
  return nowMinutes < cutoffMinutes;
}

/**
 * 실시간 가용 재고 계산 (물류센터 마감 시차 보정)
 *
 * - 쿠팡: 18시 이전이면 오늘 출고분이 아직 반영 안 됨 → "가용" = 현재고
 * - 쿠팡: 18시 이후면 오늘 출고 반영됨 → "가용" = 현재고 (이미 차감된 상태)
 * - 일반: 24시 전까지 당일 출고 반영
 *
 * "실시간 가용" = 현재 시점에서 실제로 주문 가능한 수량
 * - 물류센터 마감 전: 당일 예정 출고분을 아직 차감하지 않은 "가용" 수량
 * - 물류센터 마감 후: 이미 차감된 현재고 = 가용 수량
 *
 * @param currentStock 현재 재고 (품목별 또는 제품별)
 * @param todayOutByChannel 오늘 출고 예정/실적 (채널별)
 * @param now 기준 시각
 */
export function computeRealTimeAvailableStock(
  currentStock: number,
  todayOutByChannel: { coupang: number; general: number },
  now: Date = new Date()
): number {
  let pendingDeduction = 0;

  // 쿠팡: 18시 이전이면 오늘 쿠팡 출고분은 아직 시스템에 반영 안 됨
  // → "가용" = 현재고 - (아직 반영 안 된 쿠팡 출고)
  if (isBeforeCutoff("coupang", now)) {
    pendingDeduction += todayOutByChannel.coupang;
  }

  // 일반: 23:59 이전이면 오늘 일반 출고분 미반영
  if (isBeforeCutoff("general", now)) {
    pendingDeduction += todayOutByChannel.general;
  }

  // 실시간 가용 = 현재고 - 아직 반영되지 않은 당일 출고
  return Math.max(0, currentStock - pendingDeduction);
}

/**
 * 트랜잭션에서 오늘 출고량을 채널별로 집계 (전체)
 */
export function getTodayOutboundByChannel(
  transactions: Transaction[],
  now: Date = new Date()
): { coupang: number; general: number } {
  const today = now.toISOString().slice(0, 10);
  const result = { coupang: 0, general: 0 };

  for (const tx of transactions) {
    if (tx.type !== "out" || tx.date !== today) continue;
    const ch = (tx as TransactionWithChannel).salesChannel;
    if (ch === "coupang") result.coupang += tx.quantity;
    else result.general += tx.quantity; // 일반 또는 미지정
  }
  return result;
}

/**
 * 품목별 오늘 출고량 (채널 구분 없이 합계 - 마감 시차 보정용)
 * 채널별 보정이 필요하면 getTodayOutboundByChannelPerItem 사용
 */
export function getTodayOutboundByItem(
  transactions: Transaction[],
  now: Date = new Date()
): Record<string, { coupang: number; general: number }> {
  const today = now.toISOString().slice(0, 10);
  const result: Record<string, { coupang: number; general: number }> = {};

  for (const tx of transactions) {
    if (tx.type !== "out" || tx.date !== today) continue;
    const ch = (tx as TransactionWithChannel).salesChannel;
    const itemId = tx.itemId;
    if (!result[itemId]) result[itemId] = { coupang: 0, general: 0 };
    if (ch === "coupang") result[itemId].coupang += tx.quantity;
    else result[itemId].general += tx.quantity;
  }
  return result;
}

/**
 * 최근 N일 평균 일일 출고량 계산 (소진일 예측용)
 * @param transactions 출고 트랜잭션
 * @param itemId 품목 ID (또는 productCode로 제품별)
 * @param days 최근 일수 (기본 30일, 1년 데이터 있으면 90일 권장)
 */
export function getAverageDailyOutbound(
  transactions: Transaction[],
  itemId: string,
  days: number = 30,
  byProduct: boolean = false
): number {
  const now = new Date();
  const cutoff = new Date(now);
  cutoff.setDate(cutoff.getDate() - days);
  const cutoffStr = cutoff.toISOString().slice(0, 10);

  const dailyTotal: Record<string, number> = {};
  let totalQty = 0;

  for (const tx of transactions) {
    if (tx.type !== "out" || tx.date < cutoffStr) continue;
    const key = byProduct ? (tx.productCode ?? itemId) : itemId;
    const match = byProduct
      ? tx.productCode === itemId
      : tx.itemId === itemId;
    if (!match) continue;

    dailyTotal[tx.date] = (dailyTotal[tx.date] ?? 0) + tx.quantity;
    totalQty += tx.quantity;
  }

  const dayCount = Object.keys(dailyTotal).length;
  if (dayCount === 0) return 0;
  return totalQty / dayCount; // 실제 거래일 기준 평균
}

/**
 * 재고 소진일(Run-out date) 예측
 * @param currentStock 현재 재고 수량
 * @param avgDailyOut 평균 일일 출고량
 * @param fromDate 기준일 (기본: 오늘)
 * @returns 예상 소진일 (YYYY-MM-DD) 또는 null (소진 불가/재고 충분)
 */
export function predictRunOutDate(
  currentStock: number,
  avgDailyOut: number,
  fromDate: Date = new Date()
): { date: string | null; daysLeft: number; isInfinite: boolean } {
  if (currentStock <= 0) {
    return { date: fromDate.toISOString().slice(0, 10), daysLeft: 0, isInfinite: false };
  }
  if (avgDailyOut <= 0) {
    return { date: null, daysLeft: Infinity, isInfinite: true };
  }

  const daysLeft = Math.floor(currentStock / avgDailyOut);
  const runOut = new Date(fromDate);
  runOut.setDate(runOut.getDate() + daysLeft);

  return {
    date: runOut.toISOString().slice(0, 10),
    daysLeft,
    isInfinite: false,
  };
}

/**
 * 품목별 소진일 예측 결과
 */
export interface RunOutPrediction {
  itemId: string;
  itemName: string;
  currentStock: number;
  avgDailyOut: number;
  runOutDate: string | null;
  daysLeft: number;
  isInfinite: boolean;
  isUrgent: boolean; // 7일 이내 소진 예상
}

/**
 * 전체 품목에 대한 소진일 예측
 */
export function predictAllRunOutDates(
  stock: Record<string, number>,
  transactions: Transaction[],
  itemNames: Record<string, string>,
  lookbackDays: number = 30
): RunOutPrediction[] {
  const results: RunOutPrediction[] = [];
  const itemIds = Object.keys(stock);

  for (const itemId of itemIds) {
    const currentStock = stock[itemId] ?? 0;
    const avgDailyOut = getAverageDailyOutbound(transactions, itemId, lookbackDays, false);
    const pred = predictRunOutDate(currentStock, avgDailyOut);
    const daysLeft = pred.isInfinite ? 999 : pred.daysLeft;

    results.push({
      itemId,
      itemName: itemNames[itemId] ?? itemId,
      currentStock,
      avgDailyOut: Math.round(avgDailyOut * 10) / 10,
      runOutDate: pred.date,
      daysLeft,
      isInfinite: pred.isInfinite,
      isUrgent: !pred.isInfinite && pred.daysLeft <= 7 && pred.daysLeft >= 0,
    });
  }

  // 소진일 가까운 순 정렬 (무한대 제외)
  return results.sort((a, b) => {
    if (a.isInfinite && b.isInfinite) return 0;
    if (a.isInfinite) return 1;
    if (b.isInfinite) return -1;
    return a.daysLeft - b.daysLeft;
  });
}
